From 1f7c74b50ca4ec3a203165f8e4f40490909fa5f0 Mon Sep 17 00:00:00 2001
From: Nuri Jung <jnooree@snu.ac.kr>
Date: Tue, 22 Apr 2025 11:30:11 +0900
Subject: [PATCH 3/6] build: refactor version resolution

---
 cmake/NuriKitUtils.cmake | 84 +++++++++++++++++++++-------------------
 1 file changed, 44 insertions(+), 40 deletions(-)

diff --git a/cmake/NuriKitUtils.cmake b/cmake/NuriKitUtils.cmake
index ddd6875a..250414e3 100644
--- a/cmake/NuriKitUtils.cmake
+++ b/cmake/NuriKitUtils.cmake
@@ -3,10 +3,15 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
-macro(_nuri_get_git_version_impl)
-  find_package(Git)
+function(_nuri_get_git_version_impl)
+  set(_nuri_version_next "0.1.0.dev0")
+
+  find_package(Git QUIET)
 
   if(NOT Git_FOUND)
+    message(NOTICE "Git not found! Using ${_nuri_version_next}+unknown")
+    set(NURI_FULL_VERSION "${_nuri_version_next}+unknown" PARENT_SCOPE)
+    set(NURI_REF "unknown" PARENT_SCOPE)
     return()
   endif()
 
@@ -14,67 +19,66 @@ macro(_nuri_get_git_version_impl)
     COMMAND ${GIT_EXECUTABLE} rev-parse --short HEAD
     RESULT_VARIABLE git_result
     OUTPUT_VARIABLE nuri_revision
-    ERROR_QUIET)
+    ERROR_QUIET
+  )
   string(STRIP "${nuri_revision}" nuri_revision)
 
+  if(NOT nuri_revision)
+    set(nuri_revision "unknown")
+  endif()
+
+  execute_process(
+    COMMAND ${GIT_EXECUTABLE} symbolic-ref -q --short HEAD
+    RESULT_VARIABLE git_result
+    OUTPUT_VARIABLE nuri_branch
+    ERROR_QUIET
+  )
+  string(STRIP "${nuri_branch}" nuri_branch)
+
   execute_process(
     COMMAND ${GIT_EXECUTABLE} describe --tags --exact-match --abbrev=0
     RESULT_VARIABLE git_result
-    OUTPUT_VARIABLE NURI_REF
-    ERROR_QUIET)
+    OUTPUT_VARIABLE nuri_tag
+    ERROR_QUIET
+  )
+  string(STRIP "${nuri_tag}" nuri_tag)
 
-  if(git_result EQUAL 0)
-    string(STRIP "${NURI_REF}" NURI_REF)
+  if(nuri_tag)
+    set(NURI_REF "${nuri_tag}")
     string(REGEX REPLACE "^v" "" NURI_FULL_VERSION "${NURI_REF}")
     message(STATUS "NuriKit version from git: ${NURI_FULL_VERSION}")
   else()
-    execute_process(
-      COMMAND ${GIT_EXECUTABLE} symbolic-ref -q --short HEAD
-      RESULT_VARIABLE git_result
-      OUTPUT_VARIABLE NURI_REF
-      ERROR_QUIET)
+    set(NURI_FULL_VERSION "${_nuri_version_next}+${nuri_revision}")
+    message(NOTICE
+      "NuriKit version not found! "
+      "Assuming ${_nuri_version_next}+${nuri_revision}"
+    )
 
-    if(NOT git_result EQUAL 0)
+    if(nuri_branch)
+      set(NURI_REF "${nuri_branch}")
+    else()
       set(NURI_REF "${nuri_revision}")
     endif()
-
-    string(STRIP "${NURI_REF}" NURI_REF)
   endif()
-endmacro()
+
+  set(NURI_FULL_VERSION "${NURI_FULL_VERSION}" PARENT_SCOPE)
+  set(NURI_REF "${NURI_REF}" PARENT_SCOPE)
+endfunction()
 
 function(nuri_get_version)
   if(SKBUILD)
     # Version correctly set via scikit-build-core; skip git versioning.
     set(NURI_FULL_VERSION "${SKBUILD_PROJECT_VERSION_FULL}")
+    set(NURI_REF "v${NURI_FULL_VERSION}")
     message(
       STATUS "NuriKit version from scikit-build-core: ${NURI_FULL_VERSION}"
     )
-  else()
-    _nuri_get_git_version_impl()
-  endif()
-
-  if(NURI_FORCE_VERSION)
-    message(STATUS "Using explicit NuriKit version: ${NURI_FORCE_VERSION}")
+  elseif(NURI_FORCE_VERSION)
     set(NURI_FULL_VERSION "${NURI_FORCE_VERSION}")
-  endif()
-
-  if(NURI_REF)
-    message(STATUS "NuriKit ref from git: ${NURI_REF}")
+    set(NURI_REF "v${NURI_FULL_VERSION}")
+    message(STATUS "Using explicit NuriKit version: ${NURI_FORCE_VERSION}")
   else()
-    message(NOTICE "NuriKit ref not found! Using unknown.")
-    set(NURI_REF "unknown")
-  endif()
-
-  if(nuri_revision)
-    message(STATUS "NuriKit revision from git: ${nuri_revision}")
-  else()
-    message(NOTICE "NuriKit revision not found! Using unknown.")
-    set(nuri_revision "unknown")
-  endif()
-
-  if(NOT NURI_FULL_VERSION)
-    set(NURI_FULL_VERSION "0.1.0.dev0+${nuri_revision}")
-    message(NOTICE "NuriKit version not found! Using ${NURI_FULL_VERSION}")
+    _nuri_get_git_version_impl()
   endif()
 
   string(REGEX REPLACE "\\+.+$" "" NURI_VERSION "${NURI_FULL_VERSION}")
-- 
2.49.0

